### ISSUE TITLE
Option to specify level of compaction for literals & resources

### ISSUE THREAD

**Issue by azaroth42** [w3c/json-ld-api#33]
Merging the almost-duplicate #3 and #6, which were accepted individually as use cases.

In order to ensure a particular JSON surface structure, where entries in an array are always presented as a JSON object regardless of whether it would be unambiguous to further compact the result, it is necessary to add a new feature to declare this type of compaction is desired.

For example, a resource where only the URI is known will compact to a single string when the context declares the property as `@type: @id`. A literal that only has a value, and the language or data type are not declared, will compact to a JSON string or number.

Instead it should be possible to have the resource compact to:
   `{"@id": "uri"}`
And the literal compact to:
  `{"@value": "value"}`

**Comment by azaroth42** [comment]
Per call, my preference is a new keyword because:
 * It's less confusing than adding further exceptions to the meaning of `@type`
 * It's orthogonal (in my view) to `@type` and the other context keywords
 * Having a developer required to make multiple calls to accomplish something that could be declared seems to go against the priority of constituencies.

As a strawperson, in a context I would like to declare:

```
{
  "value": {"@id": "rdfs:value", "@compact": "@dict", "@container": "@set},
  "relation": {"@id": "dc:relation", "@type": "@id", "@compact": "@dict", "@container": "@set"}
}
```

Which would force the following input:

```
{
   "@id": "http://example.net/instance/1",
  "value": 6,
  "relation": "http://example.net/instance/2"
}
```

To compact to:

```
[
  {
    "@id": "http://example.net/instance/1",
    "http://example.org/dc/relation": [
      {
        "@id": "http://example.net/instance/2"
      }
    ],
    "http://example.org/rdfs/value": [
      {
        "@value": 6
      }
    ]
  }
]
```

Which happens to be identical to the expanded form.

**Comment by gkellogg** [comment]
As I said on the call, this is really not necessary, as the context used for compacting a document is not also used for expanding that document (unless it happens to be included in the document itself, or provided using the `expandContext` option. In this sense, the use of `@type` is unambiguous.

There are reasonable differences on what the value of `@type` should be to signal this. IMHO, `@none` is perfectly consistent with the intuitive meaning of _none_, regardless of its use elsewhere. Introducing a new keyword, such as `@dict` is not unreasonable, but not really necessary either.

This would make the context usage like the following:

```json
{
  "value": {"@id": "rdfs:value", "@type": "@none", "@container": "@set},
  "relation": {"@id": "dc:relation", "@ type": "@none", "@container": "@set"}
}
```

and would yield the same results as @azaroth42 indicated. Note that when re-expanding the document, no values would be strings if compacted using this context. We should decide what to do if they are, however, and here I would say that they be treated as simple values, but that is not significant for the use case.

**Comment by azaroth42** [comment]
@gkellogg In your approach, how do you distinguish between a string literal and a resource's URI and thus whether to use `@id` or `@value`?

e.g.

```
{
  "@context": {
    "value": {"@id": "rdfs:value", "@type": "@none", "@container": "@set"},
    "relation": {"@id": "dc:relation", "@type": "@none", "@container": "@set"}
  },
  "@id": "http://example.net/instance/1",
  "value": "http://example.net/instance/2",
  "relation": "http://example.net/instance/2"
}
```

How would the processor know what to do with this data?

**Comment by gkellogg** [comment]
Compaction always starts from the expanded form, so there's no confusion. The steps are the following:

1. Call the `compact` API.
2. Expand the input document, possibly using `expand` API.
  1.1. Expand the input document using the `http://www.w3.org/ns/json-ld#context` link relation, if present, and options possibly including `expandContext`.
3. Compact the expanded document using the provided `context`.

In the expansion phase, it presumes that the included or specified context has term definitions to make the use of string values unambiguous. When compacting, you have the term definition that prevents it from going back to a simple string representation.

It seems that you think that a document might use a context with such a term definition and have string values that should be treated differently. That's not really how the whole round-tripping was intended, so I don't see a need to complicate a term definition so it can serve multiple roles, when what we have now works as intended.

In summary, we shouldn't see a document such as you present, as it's inconsistent with the way the term definitions are specified, which could be true for many different combinations of data.

**Comment by azaroth42** [comment]
Okay, I see :) So if you had `value` with `@type: @none` in the context, then it would not be valid to have just a string, it would need to be a JSON object that gave a datatype or class. Thus setting `@type` to `@none` in the context is precisely correct -- the context is explicitly declaring that the relationship's object does not have a default interpretation and hence would require it to be in the expanded form.

:+1: from me.

**Comment by azaroth42** [comment]
Defer, pending json-ld-syntax#77 discussion.

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2018/2018-10-05-json-ld#section3-3).

- `RESOLVED:  Defer api#33 until after discussion of disambiguation of the uses of @type as rdf:type or datatype of the JSON object`
<details><summary><i class="differentiate">View the transcript</i></summary>
Rob Sanderson: link: <a href="https://github.com/w3c/json-ld-api/issues/33">https://github.com/w3c/json-ld-api/issues/33</a><br/>
Rob Sanderson: next issue, not compact urls to string<br/>
… raised in CG, suggested to use @id and @value to avoid this. @gkellog suggested @none.<br/>
… because @none would go in the context, you are simply saying that object has no type declared in context<br/>
Ivan Herman: the point is we do have this issue of using @type differently than how those of us from rdf world think of rdf:type<br/>
… we need to be very, very clear about this.<br/>
Gregg Kellogg: the way to think about @type is more like datatype<br/>
… really comes down to question of how different this concept is from original intent of @type (rdf:type)<br/>
… @type @id says treat as is URI, @type @none just says there is no datatype so you can’t compact down to a string<br/>
… when we expand, do we default to something or do we raise error?<br/>
Rob Sanderson: in the data @type is how you express rdf:type<br/>
… we need to make obvious the aliasing of @type to type and explain in the primer<br/>
… in terms of error vs. default when expanding, I would favor an error<br/>
… picking (guessing) seems dangerous<br/>
Adam Soroka: we can get into confusion about there is not a @type vs none<br/>
Gregg Kellogg: when you expand a document and type is not specified default is treat them as simple literals<br/>
… equivalent to saying default rdf:type is xsd:string<br/>
Rob Sanderson: "value": {"@type": "@none"}<br/>
Rob Sanderson: "value": "string"<br/>
Rob Sanderson: "value": {"@id": "rdf:value"}<br/>
Rob Sanderson: if we had @type @none for value then if in the data value : string, would be treated as a string<br/>
Gregg Kellogg: if in the context, would mean value is alias for string, if in data, either treated as error or treated as string<br/>
Rob Sanderson: {"@context": {"value1": {"@id": "rdf:value", "@type": "@none"}, "value2": {"@id": "rdf:value"}} …<br/>
Rob Sanderson: “value1”: “string”, “value2: “string”}<br/>
Rob Sanderson: for the above example, both would be understood the same<br/>
Gregg Kellogg: or value1 would generate an error<br/>
Benjamin Young: want to make sure rest of us aren’t just become spectators on this<br/>
Adam Soroka: so in response to @bigbluehat, let me offer my understanding<br/>
… so in rdf we distinguish between literals and strings<br/>
Simon Steyskal: -q<br/>
Adam Soroka: not the same in json-ld<br/>
… so in json-ld @type is doing several things at once<br/>
… so we keep adding more to what @type does, but it breaks down when moved into the rdf world<br/>
… so some uses of @type will add triples, some will not<br/>
… baked into the 2 different data models<br/>
… if you alias @type it aliases both for rdf:type and as datatype in the json-ld documents<br/>
… not good, but we are stuck with it, I think. So we need to document thoroughly<br/>
Rob Sanderson: seems 100% correct to me<br/>
Ivan Herman: yes, we are stuck with it, but nevertheless as a helping measure we could define a fixed alias and encouraging people to use alias differently<br/>
… down the line could become separate keys<br/>
… more than just a best practice, but also a temporizing solution<br/>
Gregg Kellogg: these issues came up in 1.0, but it’s good to mindful of the original json-ld which was find a way to understand your json as rdf<br/>
… so this is why we have context, and this is why we have to deal with URIs vs Strings<br/>
… we have discussed a default context (as a way to create default alias)<br/>
… does create an issue when compacting, because one of the aliases has to win out<br/>
… so round-tripping leads to these kinds of issues.<br/>
… we are restricted a bit on what we do for 1.1 (may be able to do more for 2.0)<br/>
Rob Sanderson: Thus {"@context": {"type": "@type", "@datatype": "@type"} ... } would allow people to use @datatype for data types, but compaction would always select one or the other for all situations<br/>
Rob Sanderson: so if we put aliases type and datatype, works for expansion, but would not roundtrip through compaction<br/>
… so maybe datatype has to more than just a context item<br/>
Gregg Kellogg: but it would break some compaction rules, so would require 1.1 being explicit<br/>
Rob Sanderson: are we ready to move forward? Or do we need more time<br/>
Ivan Herman: the latter<br/>
Jeff Mixter: I would like to mull this over more<br/>
Rob Sanderson: will make new issue about disambiguating different uses of @type<br/>
… will provide a framework for resolving issue<br/>
Gregg Kellogg: may take some time to resolve at f2f<br/>
Rob Sanderson: please add initial thoughts and then expect we may discuss in greater detail at f2f<br/>
Proposed resolution: Defer api#33 until after discussion of disambiguation of the uses of @type as rdf:type or datatype of the JSON object (Rob Sanderson)<br/>
Tim Cole: +1<br/>
Rob Sanderson: +1<br/>
Benjamin Young: +1<br/>
Ivan Herman: +1<br/>
Gregg Kellogg: +1<br/>
Adam Soroka: If LD isn’t about making the +1<br/>
Jeff Mixter: +1<br/>
Adam Soroka: +1<br/>
Simon Steyskal: +1<br/>
Resolution #3: Defer api#33 until after discussion of disambiguation of the uses of @type as rdf:type or datatype of the JSON object<br/>
</details>

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-02-08-json-ld#section5-7).

- `RESOLVED:  Use @type:@none to force compaction to generate object values ; Ivan and Adam will review for understandability {: #resolution12 .resolution}`
<details><summary><i class="differentiate">View the transcript</i></summary>
Option to specify level of compaction for literals &amp; resources<br/>
Rob Sanderson: ref: <a href="https://github.com/w3c/json-ld-api/issues/33#">https://github.com/w3c/json-ld-api/issues/33#</a><br/>
Rob Sanderson: <a href="https://github.com/w3c/json-ld-api/issues/33#issuecomment-418518683">https://github.com/w3c/json-ld-api/issues/33#issuecomment-418518683</a><br/>
Gregg Kellogg: what happens if you see data that’s a string? Answer: you won’t.<br/>
… but the alg. has to do something.<br/>
… provide a way to compact a document that does IRI compaction (aka turn IRIs into terms/compact IRIs) probably to reduce the use of arrays so you don’ t use it unless you need it.<br/>
… so we would always have objects with @id and @value. If they were aliased, it would be that, etc, etc.<br/>
… so the question that Rob raised is what if you see an input that has @value that looks like a URI, and that has @type: none. What would happen?<br/>
… this wouldn’t happen, because you’d expand that makes this unambiguous. But what id you did?<br/>
… I would say that you’d probably treat these as strings.<br/>
Rob Sanderson: I agree<br/>
Adam Soroka: It seems that it’s a little surprising, but this should be explained clearly, since it can be counterintuitive.<br/>
Gregg Kellogg: and where does this go?<br/>
… it’s not really part of the syntax.<br/>
… that matters in compaction, but to an author, I don’t think that @none has an effect, because it only matters in context evaluation<br/>
… the syntax document leaves the rules to the API document<br/>
Ivan Herman: I would welcome this into the syntax document<br/>
… people don’t read the API document.<br/>
Gregg Kellogg: but it’s not useful to understand the syntax, but only the operation of the compaction<br/>
Ivan Herman: if there are ways to control compactions, I should know them in the API document<br/>
Adam Soroka: does this go in the primer?<br/>
Gregg Kellogg: when you’re reading an document, and you see an embedded context, you get the mental model of what expansion does. Similarly, when compacting, you get the idea of how to turn an ID into a string.<br/>
… you can understand the mental model.<br/>
Ivan Herman: where is that documented?<br/>
Gregg Kellogg: there’s an incomplete description in the syntax document, and a full description would make the syntax document even MORE complicated.<br/>
Adam Soroka: Can we put a note in somewhere?<br/>
Gregg Kellogg: we need a note for it now<br/>
Rob Sanderson: let’s see what the documentation looks like<br/>
Gregg Kellogg: agreed, and someone should read through it and see what’s missing.<br/>
Adam Soroka: ajs6f: +1 to responding to something concrete<br/>
Gregg Kellogg: framing is a different, since it embodies more.<br/>
Proposed resolution: Use @type:@none to force compaction to generate object values ; Ivan and Adam will review for understandability (Rob Sanderson)<br/>
Rob Sanderson: +1<br/>
Gregg Kellogg: +1<br/>
Jeff Mixter: +1<br/>
David Newbury: +1<br/>
Ivan Herman: +1<br/>
Adam Soroka: the primer should talk a lot about shapes.<br/>
Adam Soroka: +1<br/>
David I. Lehn: +1<br/>
Harold Solbrig: +1<br/>
Resolution #12: Use @type:@none to force compaction to generate object values ; Ivan and Adam will review for understandability {: #resolution12 .resolution}<br/>
</details>

**Comment by azaroth42** [comment]
Done :)
